<a href="https://colab.research.google.com/github/John-G-Thomas/DS-Unit-1-Sprint-2-Statistics/blob/master/module2/LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [0]:
import pandas as pd
import numpy as np
from scipy import stats

In [0]:
!wget "https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip"

--2020-06-09 19:39:56--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip’

student.zip         100%[===================>]  20.00K  --.-KB/s    in 0.03s   

2020-06-09 19:39:57 (704 KB/s) - ‘student.zip’ saved [20478/20478]



In [0]:
!unzip student.zip

Archive:  student.zip
  inflating: student-mat.csv         
  inflating: student-por.csv         
  inflating: student-merge.R         
  inflating: student.txt             


In [0]:
df = pd.read_csv('student-mat.csv', sep=';')

print(df.shape)
df.head(10)

(395, 33)


school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10
5     GP   M   16       U     LE3       T  ...     2       5       10  15  15  15
6     GP   M   16       U     LE3       T  ...     1       3        0  12  12  11
7     GP   F   17       U     GT3       A  ...     1       1        6   6   5   6
8     GP   M   15       U     LE3       A  ...     1       1        0  16  18  19
9     GP   M   15       U     GT3       T  ...     1       5        0  14  15  15

[10 rows x 33 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


In [0]:
# Related Variable because p value is belove .05 (As subjects get older the less freetime they have)

In [0]:
df['age'].value_counts()

16    104
17     98
18     82
15     82
19     24
20      3
22      1
21      1
Name: age, dtype: int64

In [0]:
df['freetime'].value_counts()

3    157
4    115
2     64
5     40
1     19
Name: freetime, dtype: int64

In [0]:
contingency1 = pd.crosstab(df['age'], df['freetime'])
contingency1

freetime  1   2   3   4   5
age                        
15        3  13  35  20  11
16        5  13  47  31   8
17        5  20  32  30  11
18        4  14  36  24   4
19        2   4   7   8   3
20        0   0   0   1   2
21        0   0   0   0   1
22        0   0   0   1   0

In [0]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency1)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table", expected)

chi2 statistic 33.54250229552146
p value 0.2163427166559811
degrees of freedom 28
expected frequencies table [[ 3.9443038  13.28607595 32.59240506 23.87341772  8.30379747]
 [ 5.00253165 16.85063291 41.33670886 30.27848101 10.53164557]
 [ 4.71392405 15.87848101 38.95189873 28.53164557  9.92405063]
 [ 3.9443038  13.28607595 32.59240506 23.87341772  8.30379747]
 [ 1.15443038  3.88860759  9.53924051  6.98734177  2.43037975]
 [ 0.1443038   0.48607595  1.19240506  0.87341772  0.30379747]
 [ 0.04810127  0.16202532  0.39746835  0.29113924  0.10126582]
 [ 0.04810127  0.16202532  0.39746835  0.29113924  0.10126582]]


####Part 2

In [0]:
# Unrelated Variable because p value is above .05 meaning the two schools are similar paid by proportion 

In [0]:
df['paid'].value_counts()

no     214
yes    181
Name: paid, dtype: int64

In [0]:
df['school'].value_counts()

GP    349
MS     46
Name: school, dtype: int64

In [0]:
contingency2 = pd.crosstab(df['school'], df['paid'])
contingency2

paid     no  yes
school          
GP      188  161
MS       26   20

In [0]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency2)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table", expected)

chi2 statistic 0.03316606118405188
p value 0.8554920723086561
degrees of freedom 1
expected frequencies table [[189.07848101 159.92151899]
 [ 24.92151899  21.07848101]]


### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [0]:
df['school'].value_counts()

GP    349
MS     46
Name: school, dtype: int64

In [0]:
df['paid'].value_counts()

no     214
yes    181
Name: paid, dtype: int64

In [0]:
observe = pd.crosstab(df['school'], df['paid'])
observe

paid     no  yes
school          
GP      188  161
MS       26   20

In [0]:
observe_with_margins = pd.crosstab(df['school'], df['paid'], margins = True)
observe_with_margins

paid     no  yes  All
school               
GP      188  161  349
MS       26   20   46
All     214  181  395

In [0]:
observe = observe.values
observe

array([[188, 161],
       [ 26,  20]])

In [0]:
expected_rowl_col1 = 349*214/395

expected_rowl_col1

189.07848101265822

In [0]:
row_totals  = df['paid'].value_counts().values

row_totals

array([214, 181])

In [0]:
col_totals  = df['school'].value_counts().values

col_totals

array([349,  46])

In [0]:
sample_size = df.shape[0]

sample_size

395

In [0]:
expected = np.array([[0,0],
                     [0,0]])

for i, row in enumerate(row_totals): 
  for j, col in enumerate(col_totals):
    expected_value = (row*col / sample_size)
    expected[j][i] = expected_value

expected

array([[189, 159],
       [ 24,  21]])

In [0]:
observe

array([[188, 161],
       [ 26,  20]])

In [0]:
((observe - expected)**2 / expected)

array([[0.00529101, 0.02515723],
       [0.16666667, 0.04761905]])

In [0]:
((observe - expected)**2 / expected).sum().sum()

0.24473395228112207

In [0]:
#Conclusion: Based on  a chi2 statistic of .926, and a p-value of .3385, we fail to reject the null hypothesis that sex and internet access at home are independent

Conclusion: Based on  a chi2 statistic of 0.2447, and a p-value of 0.62081, I reject the null hypothesis that 'paid' and 'school' are Unrealated

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [0]:
x = np.array([[0,0],[0,0]])
for i, row in enumerate(row_totals): 
  for j, col in enumerate(col_totals):
    x_value = (row*col / sample_size)
    x[j][i] = x_value



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

- What are hypothesis test? How are they useful?

- What is a "null hypothesis?"

- What is a p-value and what does it represent?

- What does it mean for something to be "statistically significant?"

- What is a test statistic and how does it relate to a p-value?

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

Statistical significance is the likelihood that a relationship between two or more variables is caused by something other than chance. ... Statistical hypothesis testing is used to determine whether the result of a data set is statistically significant.

In [0]:
#Hypothesis test:Is the thing we assume and can make generizations based off of smaller samples of a larger data

In [0]:
#Null:(in a statistical test) the hypothesis that there is no significant difference between specified populations, any observed difference being due to sampling or experimental error.

In [0]:
# P-Value shows if the varibles are related and it based off of the "confidence level"

In [0]:
#Degree's of Freedom: In statistics, the number of degrees of freedom is the number of values in the final calculation of a statistic that are free to vary

 t-statistic is the ratio of the departure of the estimated value of a parameter from its hypothesized value to its standard error. It is used in hypothesis testing via Student's t-test. The T-statistic is used in a T test to determine if you should support or reject the null hypothesis.

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)